driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2201,7 +2201,14 @@ public PreparedStatement addPrepared(PreparedStatement stmt) {
 
                 // The one object in the cache will get GCed once it's not referenced by the client anymore since we use a weak reference.
                 // So we need to make sure that the instance we do return to the user is the one that is in the cache.
-                return previous;
+                if (previous.getPreparedId().getResultMetadataId().equals(stmt.getPreparedId().metadata)) {
```

Comment:
The warning line 2199 should imo only be printed if we are in this side of the if/else block.